homeassistant/components/synology_dsm/__init__.py

```diff
@@ -144,30 +247,230 @@ async def async_setup(self):
                 minutes=self._entry.options.get(
                     CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL
                 )
             ),
         )
 
+    @callback
+    def subscribe(self, api_key, unique_id):
+        """Subscribe an entity from API fetches."""
+        if api_key not in self._fetching_entities:
+            self._fetching_entities[api_key] = set()
+        self._fetching_entities[api_key].add(unique_id)
+
+        @callback
+        def unsubscribe() -> None:
+            """Unsubscribe an entity from API fetches (when disable)."""
+            self._fetching_entities[api_key].remove(unique_id)
+
+        return unsubscribe
+
+    async def _async_should_fetch_api(self):
+        """Determine if we should fetch each API, if one entity needs it."""
+        # Entities not added yet, fetch all
+        if not self._fetching_entities:
+            return
+
+        # Determine if we should fetch an API
+        self._with_security = bool(
+            self._fetching_entities.get(SynoCoreSecurity.API_KEY)
+        )
+        self._with_storage = bool(self._fetching_entities.get(SynoStorage.API_KEY))
+        self._with_utilisation = bool(
+            self._fetching_entities.get(SynoCoreUtilization.API_KEY)
+        )
+
+        # Reset not used API
+        if not self._with_security:
+            self.dsm.reset(self.security)
+            self.security = None
+
+        if not self._with_storage:
+            self.dsm.reset(self.storage)
+            self.storage = None
+
+        if not self._with_utilisation:
+            self.dsm.reset(self.utilisation)
+            self.utilisation = None
+
     def _fetch_device_configuration(self):
         """Fetch initial device config."""
         self.information = self.dsm.information
-        self.utilisation = self.dsm.utilisation
-        self.storage = self.dsm.storage
+
+        if self._with_security:
+            self.security = self.dsm.security
+
+        if self._with_storage:
+            self.storage = self.dsm.storage
+
+        if self._with_utilisation:
+            self.utilisation = self.dsm.utilisation
 
     async def async_unload(self):
         """Stop interacting with the NAS and prepare for removal from hass."""
         self._unsub_dispatcher()
 
-    async def update(self, now=None):
+    async def async_update(self, now=None):
         """Update function for updating API information."""
+        await self._async_should_fetch_api()
         await self._hass.async_add_executor_job(self.dsm.update)
         async_dispatcher_send(self._hass, self.signal_sensor_update)
+
+
+class SynologyDSMEntity(Entity):
+    """Representation of a Synology NAS entry."""
+
+    def __init__(
+        self, api: SynoApi, entity_type: str, entity_info: Dict[str, str],
+    ):
+        """Initialize the Synology DSM entity."""
+        self._api = api
+        self._api_key = entity_type.split(":")[0]
+        self.entity_type = entity_type.split(":")[-1]
```

Comment:
Since we always split this, we should pass two different variables instead.

Reply:
Not always, see `self._unique_id` line 347.

Also if we are doing this here, this avoid doing it multiple times on platforms setup.

Reply:
I'm not a fan of passing variables we control and splitting them later, but I see why you are doing it as it makes the code more reusable.